scipy/sparse/tests/test_base.py

```diff
@@ -3972,7 +3989,11 @@ def spmatrix(self, D, **kwargs):
         NC = construct(arg1, **kwargs)
 
         # check that result is valid
-        assert_allclose(NC.A, M.A)
+        if NC.dtype in [np.float32, np.complex64]:
+            rtol = 1e-04
+        else:
+            rtol = 1e-07
+        assert_allclose(NC.A, M.A, rtol=rtol)
```

Comment:
I included it because the `assert_allclose` test failed at the default rtol for `_NonCanonicalMixin` subclasses initialized with `np.random.rand(40, 40)` and `sparse.rand(40, 40, density=1e-2).A` (used in `test_sum`) for the dtypes `np.float32` and `np.complex64`.